File: src/bin/0378-kth-smallest-element-in-a-sorted-matrix.cpp
// @leetup=info id=378 lang=cpp slug=kth-smallest-element-in-a-sorted-matrix

#include <algorithm>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  int kthSmallest(std::vector<std::vector<int>>& matrix, int k) {
    std::priority_queue<std::tuple<int, size_t, size_t>> q;

    q.push({-matrix[0][0], 0, 0});
    int res;
    for (int i = 0; i < k; ++i) {
      auto [n, r, c] = q.top();
      res = -n;
      q.pop();
      if (c == 0 && r + 1 < matrix.size()) {
        q.push({-matrix[r + 1][c], r + 1, c});
      }
      if (c + 1 < matrix[0].size()) {
        q.push({-matrix[r][c + 1], r, c + 1});
      }
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0378-kth-smallest-element-in-a-sorted-matrix.py
# @leetup=info id=378 lang=python3 slug=kth-smallest-element-in-a-sorted-matrix



# @leetup=code


import heapq


class Solution:
    def kthSmallest(self, matrix: list[list[int]], k: int) -> int:
        #  1  5  9
        # 10 11 13
        # 12 13 15

        N = len(matrix)
        q = []

        heapq.heappush(q, (matrix[0][0], 0, 0))
        for _ in range(k):
            (n, r, c) = heapq.heappop(q)
            if c + 1 < N:
                heapq.heappush(q, (matrix[r][c + 1], r, c + 1))
            if c == 0 and r + 1 < N:
                heapq.heappush(q, (matrix[r + 1][c], r + 1, c))
        return n


# @leetup=code
